### PR TITLE
Support depending on potentially stale values in damlc

### PR DESCRIPTION
**Pull request by cocreature**
For now, this is opt-in and only enabled for the scenario service.
Locations should be properly mapped so if lines are inserted above a
scenario, the scenario link will move down.

Eventually, we will probably want to enable this for most rules.

It is also a precursor for completion since you usually want that on a file that does not compile at the moment.

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: compiler/damlc/tests/src/DA/Test/PositionMapping.hs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+import Test.Tasty.HUnit
+
+import Development.IDE.Core.PositionMapping
+import Language.Haskell.LSP.Types
+
+
+main :: IO ()
+main = defaultMain $
+    testGroup "position mapping"
+        [ testGroup "toCurrent"
```
Comment (cocreature) [inline compiler/damlc/tests/src/DA/Test/PositionMapping.hs]: I’m planning to extend this test suite with a bunch of property tests but I’ll leave that for a separate PR.

File: compiler/hie-core/src/Development/IDE/Core/Shake.hs [hunk 2]
```diff
@@ -382,7 +423,9 @@ instance Show k => Show (Q k) where
 
 -- | Invariant: the 'v' must be in normal form (fully evaluated).
 --   Otherwise we keep repeatedly 'rnf'ing values taken from the Shake database
-data A v = A (Value v) (Maybe BS.ByteString)
+-- Note (MK) I am not sure why we need the ShakeValue here, maybe we
```
Comment (cocreature) [inline compiler/hie-core/src/Development/IDE/Core/Shake.hs]: Just to be clear: I am not sure why we had the ByteString here before either so my confusion is not caused by this PR :slightly_smiling_face:

File: compiler/hie-core/src/Development/IDE/Core/Shake.hs [hunk 3]
```diff
@@ -245,6 +276,7 @@ shakeOpen eventer logger opts rules = do
         diagnostics <- newVar mempty
         publishedDiagnostics <- newVar mempty
         debouncer <- newDebouncer
+        positionMapping <- newVar mempty
```
Comment (hurryabit) [inline compiler/hie-core/src/Development/IDE/Core/Shake.hs]:
```suggestion
        positionMapping <- newVar Map.empty
```

File: compiler/hie-core/src/Development/IDE/Core/PositionMapping.hs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    Range <$> toCurrentPosition mapping a <*> toCurrentPosition mapping b
+
+fromCurrentRange :: PositionMapping -> Range -> Maybe Range
+fromCurrentRange mapping (Range a b) =
+    Range <$> fromCurrentPosition mapping a <*> fromCurrentPosition mapping b
+
+idMapping :: PositionMapping
+idMapping = PositionMapping Just Just
+
+applyChange :: PositionMapping -> TextDocumentContentChangeEvent -> PositionMapping
```
[the remark this answers is not in the document]
Comment (cocreature) [reply compiler/hie-core/src/Development/IDE/Core/PositionMapping.hs]: I actually had the swapped order first but it turns out the only usesite is a `foldl'` so this order is more convenient.

File: compiler/lsp-tests/src/Main.hs [hunk 5]
```diff
@@ -248,6 +249,38 @@ requestTests run _runScenarios = testGroup "requests"
                     }
               ]
 
+          closeDoc main'
+    , testCase "stale code-lenses" $ run $ do
+          main' <- openDoc' "Main.daml" damlId $ T.unlines
+              [ "daml 1.2"
+              , "module Main where"
+              , "single = scenario do"
+              , "  assert (True == True)"
```
[the remark this answers is not in the document]
Comment (cocreature) [reply compiler/lsp-tests/src/Main.hs]: Because I’m lazy and keep copying the same silly example from other tests :stuck_out_tongue:
